File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/action/BetUtils.java
package cz.fb.manaus.reactor.betting.action;

import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.collect.Ordering;
import cz.fb.manaus.core.model.Bet;
import cz.fb.manaus.core.model.BetAction;
import cz.fb.manaus.core.model.BetActionType;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkState;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

@Component
public class BetUtils {

    public List<BetAction> getCurrentActions(List<BetAction> betActions) {
        Preconditions.checkArgument(!betActions.isEmpty(), "missing bet actions");
        var lastUpdates = new LinkedList<BetAction>();
        var first = betActions.get(0);
        for (var bet : betActions) {
            validate(first, bet);
            if (bet.getBetActionType() != BetActionType.UPDATE) lastUpdates.clear();
            lastUpdates.addLast(bet);
        }
        checkState(Ordering.from(comparing(BetAction::getActionDate)).isStrictlyOrdered(lastUpdates));
        return lastUpdates;
    }

    private void validate(BetAction first, BetAction bet) {
        checkArgument(first.getPrice().getSide() == bet.getPrice().getSide());
        checkArgument(first.getSelectionId() == bet.getSelectionId());
    }

    public List<Bet> getUnknownBets(List<Bet> bets, Set<String> myBets) {
        return bets.stream().filter(bet -> !myBets.contains(bet.getBetId())).collect(toList());
    }

    public List<String> parseProposers(String proposers) {
        return Splitter.on(',').omitEmptyStrings().trimResults().splitToList(proposers);
    }

    public SettledBet limitBetAmount(double ceiling, SettledBet bet) {
        var newPrice = limitPriceAmount(ceiling, bet.getPrice());
        if (newPrice.isPresent()) {
            var copy = new SettledBet();
            BeanUtils.copyProperties(bet, copy);

            var amount = bet.getPrice().getAmount();
            var rate = ceiling / amount;
            copy.setProfitAndLoss(rate * bet.getProfitAndLoss());
            limitActionAmount(ceiling, copy);

            newPrice.ifPresent(copy::setPrice);
            return copy;
        }
        limitActionAmount(ceiling, bet);
        return bet;
    }

    private Optional<Price> limitPriceAmount(double ceiling, Price origPrice) {
        double amount = origPrice.getAmount();
        if (ceiling < amount) {
            var newPrice = new Price();
            BeanUtils.copyProperties(origPrice, newPrice);
            newPrice.setAmount(ceiling);
            return Optional.of(newPrice);
        }
        return Optional.empty();
    }

    public void limitActionAmount(double ceiling, SettledBet betCopy) {
        var orig = betCopy.getBetAction();
        if (orig != null) {
            var actionCopy = new BetAction();
            BeanUtils.copyProperties(orig, actionCopy);
            var newPrice = limitPriceAmount(ceiling, orig.getPrice());
            newPrice.ifPresent(actionCopy::setPrice);
            betCopy.setBetAction(actionCopy);
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/action/BetUtils.kt
package cz.fb.manaus.reactor.betting.action

import cz.fb.manaus.core.model.*

object BetUtils {

    fun getCurrentActions(betActions: List<BetAction>): List<BetAction> {
        validate(betActions)
        val updates = betActions.takeLastWhile { it.betActionType == BetActionType.UPDATE }
        return when (val place = betActions.getOrNull(betActions.size - updates.size - 1)) {
            null -> updates
            else -> listOf(place) + updates
        }
    }

    private fun validate(betActions: List<BetAction>) {
        val nextZip = betActions.zipWithNext()
        nextZip.forEach { check(it.first.time < it.second.time) }
        nextZip.forEach { check(it.first.selectionId == it.second.selectionId) }
        nextZip.forEach { check(it.first.price.side == it.second.price.side) }
    }

    fun getUnknownBets(bets: List<Bet>, myBets: Set<String>): List<Bet> {
        return bets.filter { it.betId !in myBets }
    }

    fun limitBetAmount(ceiling: Double, bet: RealizedBet): RealizedBet {
        val newBetPrice = limitPriceAmount(ceiling, bet.settledBet.price)
        var result = bet
        if (newBetPrice != null) {
            val amount = bet.settledBet.price.amount
            val rate = ceiling / amount
            val profitAndLoss = rate * bet.settledBet.profitAndLoss

            result = result.copy(settledBet = result.settledBet.copy(profitAndLoss = profitAndLoss))
        }
        val newActionPrice = limitPriceAmount(ceiling, bet.betAction.price)
        if (newActionPrice != null) {
            result = result.copy(betAction = result.betAction.copy(price = newActionPrice))
        }
        return result
    }

    private fun limitPriceAmount(ceiling: Double, origPrice: Price): Price? {
        val amount = origPrice.amount
        return if (ceiling < amount) origPrice.copy(amount = ceiling) else null
    }
}
